irohad/torii/processor/impl/query_processor_impl.cpp

```diff
@@ -22,15 +22,78 @@
 namespace iroha {
   namespace torii {
 
+    /**
+     * Checks if public keys are subset of given signaturies
+     * @param signatures user signatories, an iterable collection
+     * @param public_keys vector of public keys
+     * @return true if user has needed signatories, false instead
+     */
+    bool signaturesSubset(
+        const shared_model::interface::SignatureSetType &signatures,
+        const std::vector<shared_model::crypto::PublicKey> &public_keys) {
+      // TODO 09/10/17 Lebedev: simplify the subset verification IR-510
+      // #goodfirstissue
+      // TODO 30/04/2018 x3medima17: remove code duplication in query_processor
+      // IR-1192 and stateful_validator
+      std::unordered_set<std::string> txPubkeys;
+      for (auto sign : signatures) {
+        txPubkeys.insert(sign->publicKey().toString());
+      }
+      return std::all_of(public_keys.begin(),
+                         public_keys.end(),
+                         [&txPubkeys](const auto &public_key) {
+                           return txPubkeys.find(public_key.toString())
+                               != txPubkeys.end();
+                         });
+    }
+
+    /**
+     * Builds QueryResponse that contains StatefulError
+     * @param hash - original query hash
+     * @return QueryRepsonse
+     */
+    std::shared_ptr<shared_model::interface::QueryResponse> buildStatefulError(
+        const shared_model::interface::types::HashType &hash) {
+      return clone(
+          shared_model::proto::TemplateQueryResponseBuilder<>()
+              .queryHash(hash)
+              .errorQueryResponse<
+                  shared_model::interface::StatefulFailedErrorResponse>()
+              .build());
+    }
+
     QueryProcessorImpl::QueryProcessorImpl(
         std::shared_ptr<ametsuchi::Storage> storage)
         : storage_(storage) {}
 
+    bool QueryProcessorImpl::checkSignatories(
+        const shared_model::interface::Query &qry) {
+      const auto &sig = *qry.signatures().begin();
+
+      const auto &wsv_query = storage_->getWsvQuery();
+      auto qpf =
+          model::QueryProcessingFactory(wsv_query, storage_->getBlockQuery());
+      auto signatories = wsv_query->getSignatories(qry.creatorAccountId());
+      if (not signatories) {
+        subject_.get_subscriber().on_next(buildStatefulError(qry.hash()));
```

Comment:
This code is duplicated in lines 88-89 - same status response will be returned twice.